### PR TITLE
[Bug fix] Fix Quasar SDPA with fp32 dest accumulation

### LINKED ISSUE #55733: quasar sdpa: compute unpack_modes reaches the Gen1 config via std::get, throws std::bad_variant_access on Quasar
### What

`ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp:1686` reaches the compute config's `unpack_modes` by naming the Gen1 alternative of the variant returned by the arch-agnostic helper:

```cpp
auto compute_hw = ttnn::to_compute_hardware_config(device->arch(), compute_kernel_config);
if (fp32_dest_acc_en) {
    auto& gen1 = std::get<ComputeGen1Config>(compute_hw);
    gen1.unpack_modes.insert({QK_IM, tt::tt_metal::UnpackMode::UnpackToSrc});
    ...
}
```

`to_compute_hardware_config` returns whichever alternative matches `arch`, so on Quasar it returns `ComputeGen2Config` and this `std::get` throws `std::bad_variant_access`. Reached whenever `fp32_dest_acc_en` is set.

### Why this site in particular

There are other call sites with the same shape, but they are in Gen1-targeted ops and are being converted op by op as each one goes through its Metal 2.0 post-port phase (batch_norm was #55601).

This one is different on both counts:

- It lives in `experimental/quasar/`, the tree Quasar bring-up actually exercises, and it keys off the live `device->arch()` — so it is the site most likely to be reached in practice rather than latently.
- The post-port process explicitly treats `experimental/quasar/` as out of scope, so **no post-port pass will ever reach it.** #52269 covers optimizing the Quasar `unpack_modes` *values*, not the access path. Nothing currently tracks this.

### Fix

Use the generation-neutral accessor from `tt_metal/api/tt-metalium/experimental/metal2_host_api/compute_hardware_config.hpp`, which resolves whichever alternative the helper returned:

```cpp
auto& dfb_unpack_modes = unpack_modes(compute_hw);
```

No `arch` branch is needed — the helper already used `arch` to pick the alternative. On Gen1 the accessor returns the same reference to the same field, so behaviour there is unchanged. Same one-line change as #55601, which has the wording used at the converted sites.

### Owner

The file is new — it was added by #54468 (merged 2026-09-02) as a Metal 2.0 port copied into the
quasar tree, so this is active work rather than orphaned code, and that PR's author is the natural
owner. Note the production `transformer/sdpa` factory is still on the descriptor API
(`create_descriptor`, no `create_program_artifacts`), so this quasar copy is currently the only sdpa
site with this shape.

It still needs picking up explicitly, because the Metal 2.0 post-port process that would otherwise
convert it treats `experimental/quasar/` as out of scope.

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
Issue - https://github.com/tenstorrent/tt-metal/issues/55733

### PR Category

Bug fix

### Summary

This factory reached `unpack_modes` via `std::get<ComputeGen1Config>` on the variant returned by `ttnn::to_compute_hardware_config`. That helper returns the **Gen2** alternative on `ARCH::QUASAR`, so on Quasar the `std::get` cannot succeed — it throws `std::bad_variant_access`. The branch is gated on the caller-supplied `fp32_dest_acc_en` (a supported setting here), and the op is compiled and exposed to Python, so the path is reachable rather than dead.

Fixed with the generation-neutral accessor `unpack_modes(config)`, which resolves whichever alternative the helper chose. Entries unchanged; no `arch` branch needed.

Fixes #55733.

### Notes for reviewers

**No Quasar test was run for this change, on either side of it.** The failure above is established by reading the code — the variant provably cannot hold `ComputeGen1Config` on Quasar — not by reproducing it, and the fix is correspondingly unvalidated on Gen2. There is no Quasar-arch coverage for sdpa to run: the simulator path (`ARCH_NAME=quasar`) is wired up only for `binary_ng` and the TM ops.

What *is* measured is that Gen1 is unaffected: on Wormhole the op runs with `fp32_dest_acc_en=True`, and its output is bitwise identical to the pre-change build for a fixed seed (max abs diff 0.0).

The existing Gen1 tests that drive this factory with `fp32_dest_acc_en=True` — 13 across `models/experimental/llama32_1b_quasar/tests/{ops,prototype_ops,graph_ops}/` — skip on a 2-device host, since the conftest's `_CANDIDATE_REQ_SHAPES` has no entry for a system reporting `(2,1)`. They want an N150 `(1,1)`, N300 `(1,2)`, T3K `(2,4)` or Galaxy `(8,4)`.

`gen1` is renamed to `dfb_unpack_modes` because it no longer refers to a Gen1 config.

### Pipeline status

- [x] Sanity
- [x] (Runtime) Sanity Test
- [x] Nightly for sdpa Category

---

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/sdpa_quasar)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/sdpa_quasar)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/sdpa_quasar)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/sdpa_quasar)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/sdpa_quasar)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/sdpa_quasar)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/sdpa_quasar)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp [hunk 1]
```diff
@@ -1683,10 +1683,14 @@ ttnn::device_operation::ProgramArtifacts SDPAOperation::SDPAProgramFactory::crea
...
         // UnpackToSrc (no explicit unpack_to_dest_mode), preserved here.
-        auto& gen1 = std::get<ComputeGen1Config>(compute_hw);
-        gen1.unpack_modes.insert({QK_IM, tt::tt_metal::UnpackMode::UnpackToSrc});
-        gen1.unpack_modes.insert({SUM_A, tt::tt_metal::UnpackMode::UnpackToSrc});
-        gen1.unpack_modes.insert({SUM_B, tt::tt_metal::UnpackMode::UnpackToSrc});
+        // Reach unpack_modes through the generation-neutral accessor rather than
+        // std::get<ComputeGen1Config>: the helper above builds the config from device->arch(), so on
+        // Quasar it returns the Gen2 alternative and naming Gen1 here throws std::bad_variant_access.
+        // TODO(#52269): Quasar unpack_modes are copied from Gen1 and not yet optimized for Quasar.
+        auto& dfb_unpack_modes = unpack_modes(compute_hw);
```
[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp:1690]: Not adding one here.

Quasar coverage in this repo runs under the ttsim simulator, and today it exists only for `binary_ng` and the TM ops — see the run instructions in the `test_binary_ng_no_bcast.py` docstring and the `_on_quasar()` env detection in `binary_ng_quasar_test_utils.py`. There is no Quasar harness for sdpa: #54468, which added this factory, added none either.

Agreed on the underlying point though — the Wormhole comparison in the description shows no Gen1 regression, it does not guard against this site reverting. Tracked in #55733.

[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp:1690]: Agreed — retitled to `[Bug fix] Fix Quasar SDPA failure with fp32 dest accumulation`.

You're right that the old one led with the fix mechanism. Checked the convention on `main` and it backs this: recent bug-fix titles state the defect or its impact (`Fix TP-sharded KV ring indexer hang`, `Fix ttnn.sort duplicating indices for tied values`, `Honour memory_config in quasar outer`) — none names the API used to fix it.

One wording change from your suggestion: "failure" rather than "crash", since it is a `std::bad_variant_access` that surfaces as a catchable error at op launch, not a segfault. The accessor detail stays in the body where it is useful to reviewers.

**Review by Aswinmcw — APPROVED** [review]
LGTM. Verified the accessor resolves both alternatives (`unpack_modes(ComputeHardwareConfig&)` visits the variant), that this is the only `std::get<ComputeGen1Config>` under `experimental/quasar/`, and that no other Gen1-specific field access is left in the file.

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp:1690]: Correction to my earlier reply — I said there is no Quasar harness for sdpa and that #54468 added none. That was wrong.

`models/experimental/llama32_1b_quasar/tests/` has per-op tests that call `ttnn.experimental.quasar.transformer.scaled_dot_product_attention`, and #54468 updated several of them. `ops/test_scaled_dot_product_attention.py:68` even sets `fp32_dest_acc_en=True`, so it does drive the branch changed here — 13 tests across `ops/`, `prototype_ops/` and `graph_ops/` route through this factory (chunked included; decode/paged/joint are separate factories).

What holds is the narrower point behind your comment: those tests run on **Wormhole**, where both the old and new code work, so they would not catch this site reverting to `std::get<ComputeGen1Config>`. A Gen2-arch test is what would, and the simulator path (`ARCH_NAME=quasar`) is currently wired up only for `binary_ng` and the TM ops. That gap is real and stays tracked in #55733.

Note also that those 13 skip on a 2-device system: the conftest's `_CANDIDATE_REQ_SHAPES` has no entry for a system reporting `(2, 1)`, so they need an N150 `(1,1)`, N300 `(1,2)`, T3K `(2,4)` or Galaxy `(8,4)` host.

**Review by gajanan-choudhary — APPROVED** [review]
LGTM.

Should we be doing this for all Metal 2.0 ports by the way? I believe many ports might be using `std::get<ComputeGen1Config>(compute_hw);` or some equivalent to get Gen1.

**Comment by VirdhatchaniKN** [comment]
> LGTM.
> 
> Should we be doing this for all Metal 2.0 ports by the way? I believe many ports might be using `std::get<ComputeGen1Config>(compute_hw);` or some equivalent to get Gen1.

Yes @gajanan-choudhary , I beleive. it's already one of the Metal 2.0 post-port passes, run per op. batch_norm went through it in #55601.

There are 20 such sites on main today. Three are covered by the two open PRs, leaving 16 unguarded: 7 in `normalization/softmax`, 5 in `moreh_softmax`, 3 in `moreh_norm/ord_other`, 1 in `moreh_matmul`. One more is different in kind — `layernorm_distributed_metal2_helpers.hpp:78` wraps the `std::get` in a `TT_FATAL` on `holds_alternative`, so it refuses with a clear message instead of throwing. Still wants converting, but that also means retiring the guard.

This sdpa one needed its own PR only because post-port passes skip `experimental/quasar/`, so nothing would have reached it.

A sweep PR would work technically

**Review by bbradelTT — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]
